packages/wdio-browserstack-service/src/insights-handler.ts

```diff
@@ -0,0 +1,458 @@
+        const gherkinDocument = world.gherkinDocument
+        const featureData = gherkinDocument.feature
+
+        const uniqueId = getUniqueIdentifierForCucumber(world)
+
+        const testMetaData: TestMeta = {
+            uuid: uuidv4(),
+            startedAt: (new Date()).toISOString()
+        }
+
+        if (pickleData) {
+            testMetaData.scenario = {
+                name: pickleData.name,
+            }
+        }
+
+        if (gherkinDocument && featureData) {
+            testMetaData.feature = {
+                path: gherkinDocument.uri,
+                name: featureData.name,
+                description: featureData.description,
+            }
+        }
+
+        this._tests[uniqueId] = testMetaData
+
+        await this.sendTestRunEventForCucumber(world, 'TestRunStarted')
+    }
+
+    async afterScenario (world: ITestCaseHookParameter) {
+        await this.sendTestRunEventForCucumber(world, 'TestRunFinished')
+    }
+
+    async beforeStep (step: Frameworks.PickleStep, scenario: Pickle) {
+        const uniqueId = getUniqueIdentifierForCucumber({ pickle: scenario } as ITestCaseHookParameter)
+        let testMetaData = this._tests[uniqueId]
+        if (!testMetaData) {
+            testMetaData = {
+                steps: []
+            }
+        }
+
+        if (testMetaData && !testMetaData.steps) {
+            testMetaData.steps = []
+        }
+
+        testMetaData.steps?.push({
+            id: step.id,
+            text: step.text,
+            keyword: step.keyword,
+            started_at: (new Date()).toISOString()
+        })
+
+        this._tests[uniqueId] = testMetaData
+    }
+
+    async afterStep (step: Frameworks.PickleStep, scenario: Pickle, result: Frameworks.PickleResult) {
+        const uniqueId = getUniqueIdentifierForCucumber({ pickle: scenario } as ITestCaseHookParameter)
+        let testMetaData = this._tests[uniqueId]
+        if (!testMetaData) {
+            testMetaData = {
+                steps: []
+            }
+        }
+
+        if (!testMetaData.steps) {
+            testMetaData.steps = [{
+                id: step.id,
+                text: step.text,
+                keyword: step.keyword,
+                finished_at: (new Date()).toISOString(),
+                result: result.passed ? 'PASSED' : 'FAILED',
+                duration: result.duration,
+                failure: result.error ? removeAnsiColors(result.error) : result.error
+            }]
+        }
+        const stepDetails = testMetaData.steps?.find(item => item.id === step.id)
+        if (stepDetails) {
+            stepDetails.finished_at = (new Date()).toISOString()
+            stepDetails.result = result.passed ? 'PASSED' : 'FAILED'
+            stepDetails.duration = result.duration
+            stepDetails.failure = result.error ? removeAnsiColors(result.error) : result.error
+        }
+
+        this._tests[uniqueId] = testMetaData
+    }
+
+    //@ts-ignore
```

Comment:
I added this as I was getting this error due to recursive nature of the method.
`error TS7023: 'uploadPending' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions.`

Can you suggest how to solve this?

Reply:
define a return type explicitly:
```ts
    async uploadPending (
        waitTimeout = DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS,
        waitInterval = DEFAULT_WAIT_INTERVAL_FOR_PENDING_UPLOADS
    ): Promise<never> {
```